### PR TITLE
DX-64328 Array types for Gandiva

### PR DESCRIPTION
**Pull request by lriggs**
Add List input and output types for Gandiva functions. Add new reference implementations for array_contains and array_remove, tested via integration with Dremio. int32, int64, double and float list types have been tested.

* Support List types in function specification and llvm code generation.
* Pass back function type information through the expression registry.

See 1p here: https://docs.google.com/document/d/1exwXdUUnk5FqZLzVZyTdhqgwxTk0u9bL54aLVNM5Tas/edit

### CODE REVIEW

File: java/gandiva/src/main/java/org/apache/arrow/gandiva/evaluator/ListVectorExpander.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+   * @param toCapacity the size to which the buffer should be expanded to.
+   *
+   * @return address and size  of the buffer after expansion.
+   */
+  public ExpandResult expandOutputVectorAtIndex(int index, long toCapacity) {
+    if (index >= bufferVectors.length || bufferVectors[index] == null) {
+      throw new IllegalArgumentException("invalid index " + index);
+    }
+
+    int valueBufferIndex = 1;
```
Comment (DenisTarasyuk) [inline java/gandiva/src/main/java/org/apache/arrow/gandiva/evaluator/ListVectorExpander.java]: These two should be moved to constants.

File: java/gandiva/src/main/java/org/apache/arrow/gandiva/evaluator/Projector.java [hunk 2]
```diff
@@ -352,18 +351,20 @@ private void evaluate(int numRows, List<ArrowBuf> buffers, List<ArrowBuffer> buf
...
     idx = 0;
     int outColumnIdx = 0;
     for (ValueVector valueVector : outColumns) {
-      boolean isFixedWith = valueVector instanceof FixedWidthVector;
-      boolean isVarWidth = valueVector instanceof VariableWidthVector;
-      if (!isFixedWith && !isVarWidth) {
-        throw new UnsupportedTypeException(
-            "Unsupported value vector type " + valueVector.getField().getFieldType());
+      if (valueVector instanceof ListVector) {
+        outAddrs = new long[5 * outColumns.size()];
```
Comment (DenisTarasyuk) [inline java/gandiva/src/main/java/org/apache/arrow/gandiva/evaluator/Projector.java]: Move 5 to constant

File: java/gandiva/src/main/java/org/apache/arrow/gandiva/evaluator/Projector.java [hunk 3]
```diff
@@ -374,19 +375,41 @@ private void evaluate(int numRows, List<ArrowBuf> buffers, List<ArrowBuffer> buf
...
+        hasVariableWidthColumns = true;
+        resizableListVectors[outColumnIdx] = (ListVector) valueVector;
+        List<ArrowBuf> fieldBufs = ((ListVector) valueVector).getDataVector().getFieldBuffers();
+        outAddrs[idx] = valueVector.getOffsetBuffer().memoryAddress();
+        outSizes[idx++] = valueVector.getOffsetBuffer().capacity();
+
+        //vector valid
+        outAddrs[idx] = ((ListVector) valueVector).getDataVector().getFieldBuffers().get(0).memoryAddress();
+        outSizes[idx++] = ((ListVector) valueVector).getDataVector().getFieldBuffers().get(0).capacity();
+
```
Comment (DenisTarasyuk) [inline java/gandiva/src/main/java/org/apache/arrow/gandiva/evaluator/Projector.java:391]: 0 and 1 should use same constants as in ListVectorExpander

File: cpp/src/gandiva/annotator.cc [hunk 4]
```diff
@@ -80,17 +90,73 @@ void Annotator::PrepareBuffersForField(const FieldDescriptor& desc,
...
+        
+      } else {
+        // if list field is input field, just put buffer data into eval batch
+        uint8_t* child_offsets_buf = const_cast<uint8_t*>(
+            array_data.child_data.at(0)->buffers[buffer_idx]->data());
+        eval_batch->SetBuffer(desc.child_data_offsets_idx(), child_offsets_buf,
+                              array_data.child_data.at(0)->offset);
+
+        uint8_t* child_valid_buf = const_cast<uint8_t*>(
+            array_data.child_data.at(0)->buffers[0]->data());
```
Comment (DenisTarasyuk) [inline cpp/src/gandiva/annotator.cc:118]: Make 0 index a constant

File: cpp/src/gandiva/array_ops.cc [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+{
+  std::vector<Type> newInts;
+
+  const int32_t* entry_validityAdjusted = entry_validity - (loop_var );
+  int64_t validityBitIndex = 0;
+  //The validity index already has the current row length added to it, so decrement.
+  validityBitIndex = validity_index_var - entry_len;
+  std::vector<bool> outValid;
+  for (int i = 0; i < entry_len; i++) {
+    Type entry_item = *(entry_buf + (i * 1));
```
Comment (DenisTarasyuk) [inline cpp/src/gandiva/array_ops.cc]: why multiply by 1?

File: cpp/src/gandiva/projector.cc [hunk 6]
```diff
@@ -217,12 +220,40 @@ Status Projector::Evaluate(const arrow::RecordBatch& batch,
...
+      if (arrow::is_binary_like(child_data->type->id())) {
+        /* when allocate array data, child data length is an initialized value,
+         * after calculating, child data offsets buffer has been resized for results,
+         * but array data length is unchanged.
+         * We should recalculate child data length and make ArrayData with new length
+         *
+         * Otherwise, child data offsets buffer length is data length + 1
+         * and offset data is int32_t, need use buffer->size()/4 - 1
+         */
+        child_data_size = child_data->buffers[1]->size() / 4 - 1;
```
Comment (DenisTarasyuk) [inline cpp/src/gandiva/projector.cc]: 4 and 8 should be constants
1 for buffer index too

File: cpp/src/gandiva/projector.cc [hunk 7]
```diff
@@ -263,7 +313,27 @@ Status Projector::AllocArrayData(const DataTypePtr& type, int64_t num_records,
...
+    auto internal_type = type->field(0)->type();
+    ArrayDataPtr child_data;
+    if (arrow::is_primitive(internal_type->id())) {
+      child_data = arrow::ArrayData::Make(internal_type, 0 /*initialize length*/,
+                                          {std::move(data_valid_buffer), std::move(buffers[2])}, 0);
+    }
+    if (arrow::is_binary_like(internal_type->id())) {
+      child_data = arrow::ArrayData::Make(
+          internal_type, 0 /*initialize length*/,
+          {nullptr, std::move(buffers[2]), std::move(buffers[3])}, 0);
```
Comment (DenisTarasyuk) [inline cpp/src/gandiva/projector.cc:331]: 0,1,2,3 indexes as constants

File: cpp/src/gandiva/tests/list_test.cc [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+  std::cout << "Check results" << std::endl;
+  EXPECT_ARROW_ARRAY_EQUALS(array, outputs[0]);
+  // EXPECT_ARROW_ARRAY_EQUALS will not check the length of child data, but
+  // ArrayData::Slice method will check length. ArrayData::ToString method will call
+  // ArrayData::Slice method
+  EXPECT_TRUE(array->ToString() == outputs[0]->ToString());
+  EXPECT_TRUE(array->null_count() == outputs[0]->null_count());
+}
+
+/*
```
Comment (DenisTarasyuk) [inline cpp/src/gandiva/tests/list_test.cc:135]: Do you need this commented code?

File: java/gandiva/src/main/cpp/jni_common.cc [hunk 9]
```diff
@@ -625,8 +682,61 @@ Status make_record_batch_with_buf_addrs(SchemaPtr schema, int num_rows,
...
+      jlong offsets_size = in_buf_sizes[sz_idx++];
+
+      auto child_offsets_buffer = std::shared_ptr<arrow::Buffer>(  new arrow::Buffer(reinterpret_cast<uint8_t*>(offsets_addr), offsets_size));
+    
+      buffers.push_back(std::move(child_offsets_buffer));
+    }
+  }
+
+  if (type->id() == arrow::Type::LIST) {
+    jlong offsets_addr = in_buf_addrs[buf_idx++];
```
Comment (DenisTarasyuk) [inline java/gandiva/src/main/cpp/jni_common.cc:716]: Does gandiva tests perform any memory leak checks? How do we make sure that added code does not result in memory leaks?

File: java/gandiva/src/main/cpp/jni_common.cc [hunk 10]
```diff
@@ -625,8 +682,61 @@ Status make_record_batch_with_buf_addrs(SchemaPtr schema, int num_rows,
...
+    jlong offsets_size = in_buf_sizes[sz_idx++];
+    auto data_buffer = std::shared_ptr<arrow::Buffer>(  new arrow::Buffer(reinterpret_cast<uint8_t*>(offsets_addr), offsets_size));
+    auto internal_type = type->field(0)->type();
+    std::shared_ptr<arrow::ArrayData> child_data;
+    if (arrow::is_primitive(internal_type->id())) {
+      child_data = arrow::ArrayData::Make(internal_type, 0,
+                                          {std::move(buffers[2]), std::move(data_buffer)});
+    }
+    if (arrow::is_binary_like(internal_type->id())) {
```
Comment (DenisTarasyuk) [inline java/gandiva/src/main/cpp/jni_common.cc:725]: Is this block required?

File: cpp/src/gandiva/tests/list_test.cc (hunk 8 above)
Comment (lriggs) [reply cpp/src/gandiva/tests/list_test.cc:135]: I was hoping to fix it

**Pull request merged**